0000-modules/detailed-design/loading-files.md

```diff
@@ -0,0 +1,231 @@
+# Loading Files
+
+When building a Rust project, rustc will load and parse some files as Rust code
+in addition to the root module. These will be used to construct a module tree.
+By default, cargo will generate the list of files to load in this way for you,
+though you can generate such a list yourself and specify it in your
+`Cargo.toml`, or you can generate the list in another way for your non-cargo
+build system.
+
+This eliminates the need to write `mod` statements to add new files to your
+project. Instead, files will be picked up automatically as a part of your
+module hierarchy.
+
+## Detailed design
+
+### Processing the `--modules` list (rustc)
+
+rustc takes a new argument called `modules`, which takes a space separated list
+of files. Each file will be treated as a module, and rustc will attempt to open
+and parse every file listed, reporting errors if it is unable to. It will mount
+these files as a tree of Rust modules using rules which mirror the current
+rules for looking up modules.
+
+It will not attempt to open or parse files where:
+
+* The file name is not a valid Rust identifier followed by `.rs`.
+* The file is not in a subdirectory of the directory containing the root
+module.
+* Any of the subdirectories of the root module in the path to this file are not
+valid Rust identifiers.
+
+Cargo's default system will not pass any files that would be ignored by these
+conditions, but if they are passed by some other system, they are ignored
+regardless. For example, in a cargo managed crate with no dependencies, this
+would be a valid way to invoke rustc by hand:
+
+```
+rustc src/lib.rs --modules src/*.rs src/**/*.rs
```

Comment:
Nit: For clarity, I think this should either be `--modules "src/*.rs src/**/*.rs"` or `--modules src/*.rs --modules src/**/*.rs` (otherwise `src/**/*.rs` would be a "free" argument).